Support EIP-7702 Delegations in Forge

evchip opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests.
-->

## Motivation
Adds initial support for EIP-7702 transaction delegation in Forge scripts. 

Addresses #7128
<!--
Explain the context and why you're making that change. What is the problem
you're trying to solve? In some cases there is not a problem and this can be
thought of as being the motivation for your change.
-->

## Solution
* adds EIP-7702 support to forge scripts for transaction batching + delegation.
* new cheatcodes:

`createDelegation`: generates auth hash
`signDelegation`: signs auth with pk
`attachDelegation`: connects signed delegation to sender

## Limitations:
* clunky sig handling - requires passing v,r,s separately bc passing raw sig bytes breaks tests
* only supports raw private keys, needs hardware wallet etc support

## Questions:

1. best way to propagate delegation state from cheatcodes -> broadcast?
2. should we track delegations in ScriptSequence instead of ScriptResult?

looking for feedback on state handling and current api limitations before expanding functionality.

zerosnacks commented on crates/cheatcodes/src/script.rs:
```diff
@@ -2,9 +2,13 @@
+use alloy_sol_types::SolValue;
 use foundry_wallets::{multi_wallet::MultiWallet, WalletSigner};
 use parking_lot::Mutex;
+use revm::primitives::SignedAuthorization;
```
Tip: this is available in the `alloy_eips::eip7702::SignedAuthorization;` crate

klkvr commented on crates/cheatcodes/src/script.rs:
```diff
@@ -174,6 +236,7 @@ fn broadcast(ccx: &mut CheatsCtxt, new_origin: Option<&Address>, single_call: bo
         original_origin: ccx.ecx.env.tx.caller,
         depth: ccx.ecx.journaled_state.depth(),
         single_call,
+        delegation,
```
hmm, why do we need to keep track of delegation here?

evchip replied:
Removed, thanks

klkvr commented on testdata/default/cheats/AttachDelegation.t.sol:
```diff
@@ -0,0 +1,20 @@
+        bytes32 delegation = vm.createDelegation(implementation, nonce);
+        (uint8 v, bytes32 r, bytes32 s) = vm.signDelegation(delegation, pk);
+
+        vm.attachDelegation(implementation, nonce, v, r, s);
```
I believe delegations would be attached to separate transactions vs broadcast blocks, so it would be something like 
```solidity
vm.startBroadcast();
vm.attachDelegation(implementation, nonce, v, r, s);
// this transactions would be broadcasted along with attached 7702 delegation
token.transfer(....);
```

evchip replied:
thanks for the comment! looking at this closer - since 7702 makes the EOA itself the target of the tx, i think we need to route all calls through the EOA first. 

right now we can do this with a raw call:
```solidity
vm.startBroadcast();
vm.attachDelegation(implementation, nonce, v, r, s);
alice.call(abi.encodeCall(implementation.execute, (calls)));
```
but for better UX, maybe we could either:

1. change attachDelegation to take authority instead of implementation (since we have the implementation in the call itself):
```solidity
vm.attachDelegation(alice, nonce, v, r, s);
implementation.execute(calls);
```
2. or add a helper like:
```solidity
vm.delegatedCall(alice, implementation.execute, calls);
```
what do you think would be more ergonomic for users?

klkvr commented on crates/cheatcodes/src/script.rs:
```diff
@@ -29,6 +32,57 @@ impl Cheatcode for broadcast_2Call {
+}
+
+impl Cheatcode for signDelegationCall {
+    fn apply_stateful(&self, _: &mut CheatsCtxt) -> Result {
```
isn't this same as `vm.sign`? I think we could add a cheat which would both create and sign delegation in a single call

eventually once we have hardware wallets support we could also enable this cheat for more wallets than just raw keys

evchip replied:
Good call, I've combined `createDelegation` into `signDelegation`.

klkvr commented:
> best way to propagate delegation state from cheatcodes -> broadcast?

We should attach them to `TransactionRequest` here:  https://github.com/foundry-rs/foundry/blob/17e0981a071fbd3b5a0a59affb4d638a28dfec89/crates/cheatcodes/src/inspector.rs#L1004-L1016

> should we track delegations in ScriptSequence instead of ScriptResult?
I think we need to track them in both. In `ScriptResult` as part of `BroadcastableTransactions` and in `ScriptSequence` as `TransactionWithMetadata`

evchip commented:
@klkvr thanks for the detailed feedback! Will push an update soon that addresses your comments.

evchip commented:
@klkvr thanks for the pointer on `TransactionRequest`. i've added the `authorization_list` field and i'm constructing 7702 txs (type=0x04) when that field is populated. however, the delegated calls don't seem to be executing properly - i suspect we also need to write the delegation designation (0xef0100 || address) to the EOA's code somewhere. 

looking at the anvil tests, it seems revm supports 7702, but i'm not clear on where in foundry we should be handling the delegation setup. any guidance? thanks in advance.

klkvr commented:
> @klkvr thanks for the pointer on `TransactionRequest`. i've added the `authorization_list` field and i'm constructing 7702 txs (type=0x04) when that field is populated. however, the delegated calls don't seem to be executing properly - i suspect we also need to write the delegation designation (0xef0100 || address) to the EOA's code somewhere.
> 
> looking at the anvil tests, it seems revm supports 7702, but i'm not clear on where in foundry we should be handling the delegation setup. any guidance? thanks in advance.

right, we need to execute something similar to this https://github.com/bluealloy/revm/blob/346aa01398aa9b4e4e36776c21f6fe7169ed5d87/crates/revm/src/handler/mainnet/pre_execution.rs#L102 in `call` hook where we push to `broadcastable_transactions`

we can probably skip some of the steps as we will anyway execute this later during on-chain simulation, but I think nonce check and increase should be kept because those may affect contract deployment addresses

also we'll need to respect the authorization list here https://github.com/foundry-rs/foundry/blob/d2ed15d517a3af56fced592aa4a21df0293710c5/crates/script/src/simulate.rs#L120 when configuring `TxEnv` for simulation, this might require small changes to executor

evchip commented:
Hey @klkvr - got an initial implementation working but took a slightly different approach than suggested. instead of handling the delegation in the call hook, i'm writing the delegation code directly when vm.attachDelegation is called, and letting normal call routing handle delegation during simulation.

seems to work but would appreciate your eyes on it in case i'm missing edge cases. main changes are in new `attachDelegation` cheatcode. The forge test shows delegation working with gas sponsorship.

lmk what you think! happy to adjust the approach if needed.

evchip commented:
@klkvr I've corrected some faulty nonce comparison logic in the `attachDelegation` cheatcode logic. Nonce is still implicitly checked because if user's nonce is incremented in between delegation signing and execution, the authority derived from the signature will not match the authority param. I've added several tests to ensure correct validation of the nonce and signature. Looking forward to your feedback. Thanks!

evchip commented:
> @evchip it looks like current implementation doesn't actually attach delegation to resulting script transaction?

@klkvr you're right, thanks for checking. I've pushed logic to ensure that delegations are attached for script transactions. I've been testing with a local forge script, which works with the hardcoded values you see here:

https://github.com/evchip/foundry/blob/feature/forge-eip-7702/crates/cheatcodes/src/inspector.rs#L1029C25-L1031C99

However, I'm not clear on how I can set dynamic `gas` and `max_priority_fee_per_gas` at this stage. Any pointers would be greatly appreciated.

On another note, https://odyssey-explorer.ithaca.xyz does not show EIP 7702 txs at this point. I can only find them via cast: `cast receipt 0xacd9af3d4514f5ca471fd3927e2cc448b23781ebe256eb4b717c60b0911bca77 --rpc-url https://odyssey.ithaca.xyz`

evchip commented:
@klkvr I've made two changes in [this commit](https://github.com/foundry-rs/foundry/pull/9236/commits/a833dd6468b158bf15c06b7c1d47ffe88a30f413):

1. Removed hardcoded gas values for 7702 txs and replaced them with proper dynamic gas estimation that accounts for both execution cost and delegation overhead costs defined in EIP-7702 (25k gas per delegated account)

2. Added gas estimation in two places:
   - During simulation in `ScriptTransactionBuilder::with_execution_result`
   - During broadcast in `estimate_gas` for chains that require RPC gas estimation

Both changes ensure EIP-7702 transactions get sufficient gas to cover both:
- the actual execution cost (estimated via simulation/RPC)
- the EIP-7702 delegation overhead (25k per delegated account)

This handles the issue we saw where removing hardcoded values led to insufficient gas estimation. The gas estimation now properly scales with transaction complexity while maintaining the minimum required overhead for delegation.

Let me know if you want me to add tests specifically for the 7702 gas estimation logic. Thanks

klkvr commented on crates/cheatcodes/src/inspector.rs:
```diff
@@ -1001,18 +1014,27 @@ where {
+                        self.active_delegation.and_then(|addr| self.delegations.remove(&addr))
+                    {
+                        tx_req.authorization_list = Some(vec![auth_list]);
+                        tx_req.transaction_type = Some(4); // EIP-7702
```
there's no need to set tx type here I believe it would be correctly resolved on its own when building tx envelope

evchip replied:
Removed, thanks.

klkvr commented on crates/script/src/transaction.rs:
```diff
@@ -159,8 +160,13 @@ impl ScriptTransactionBuilder {
         if !self.transaction.is_fixed_gas_limit {
             if let Some(unsigned) = self.transaction.transaction.as_unsigned_mut() {
+                let mut gas = result.gas_used;
+                // if this is a 7702 tx, add the per-account auth costs
+                if let Some(auth_list) = unsigned.authorization_list.as_ref() {
+                    gas += auth_list.len() as u64 * PER_EMPTY_ACCOUNT_COST;
+                }
                 // We inflate the gas used by the user specified percentage
-                unsigned.gas = Some(result.gas_used * gas_estimate_multiplier / 100);
+                unsigned.gas = Some(gas * gas_estimate_multiplier / 100);
```
looks like this is required because we never set the auth list when simulating tx here https://github.com/foundry-rs/foundry/blob/c4a31a624874ab36284fca4e48d2197e43a62fbe/crates/script/src/simulate.rs#L121

if we set it correctly the gas used would contain the usage for the auth list as well removing the need to handle it manually

evchip replied:
Thanks - you were right about letting revm handle the 7702 gas. I've removed the manual gas calc and properly set the authorization list in the tx env before simulation.

evchip commented on crates/script/src/broadcast.rs:
```diff
@@ -41,11 +41,18 @@ where
-    );
+            100;
+
+    tx.set_gas_limit(execution_cost + auth_list_cost);
```
Removed and fixed by setting auth list on the tx env in simulation as you suggested. Thanks.

klkvr commented on crates/script/src/simulate.rs:
```diff
@@ -112,10 +112,16 @@ impl PreSimulationState {
-            .map(|transaction| async {
+            .map(|mut transaction| async {
                 let mut runner = runners.get(&transaction.rpc).expect("invalid rpc url").write();
+                let tx = transaction.tx_mut();
+
+                if let Some(tx) = tx.as_unsigned_mut() {
+                    if let Some(authorization_list) = &tx.authorization_list {
+                        runner.executor.set_delegation(authorization_list)?;
+                    }
```
let's add a helper here to get the `authorization_list` https://github.com/foundry-rs/foundry/blob/c4a31a624874ab36284fca4e48d2197e43a62fbe/crates/common/src/transactions.rs#L223

signed transactions might have one as well

evchip replied:
Fixed, thanks

klkvr commented on crates/cheatcodes/src/inspector.rs:
```diff
@@ -373,6 +376,14 @@ pub struct Cheatcodes {
+    /// When a delegation is attached via `vm.attachDelegation()`, this address is set and used
+    /// to look up the authorization data from the `delegations` map when broadcasting
+    /// transactions.
+    pub active_delegation: Option<Address>,
```
can we just have `Option<SignedAuthorization>` here which would be set on `attachDelegation` and `.take()`'n when attaching to the script tx?

evchip replied:
Fixed, thanks

klkvr commented on crates/cheatcodes/src/script.rs:
```diff
@@ -29,6 +32,68 @@ impl Cheatcode for broadcast_2Call {
+        }
+
+        // write delegation code
+        let bytecode = Bytecode::new_eip7702(*implementation);
+        authority_acc.info.code = Some(bytecode.clone());
+        authority_acc.info.code_hash = bytecode.hash_slow();
+        authority_acc.mark_touch();
```
there's `set_code` helper on `journaled_state` we can use, check out the `etch` implementation:
https://github.com/foundry-rs/foundry/blob/c4a31a624874ab36284fca4e48d2197e43a62fbe/crates/cheatcodes/src/evm.rs#L442-L451

klkvr replied:
also it looks like we don't handle nonce here at all? we should account for delegation increasing the nonce

evchip replied:
Fixed both, thanks

klkvr commented on crates/script/src/simulate.rs:
```diff
@@ -112,10 +112,14 @@ impl PreSimulationState {
                 let mut runner = runners.get(&transaction.rpc).expect("invalid rpc url").write();
+                let tx = transaction.tx_mut();
+
+                if let Some(authorization_list) = tx.authorization_list() {
+                    runner.executor.set_delegation(&authorization_list)?;
```
looks like this `set_delegation` call would set the authorization list not just for this transaction but for all transactions after it as well?

evchip replied:
Based on my tests I think the current logic will use the delegation when the EOA is called, but not on calls to any other accounts
```
vm.attachDelegation(impl, alice, v, r, s);
SimpleDelegateContract(alice).execute(calls1);
// works the same as call1 because we have not changed/removed the delegation
SimpleDelegateContract(alice).execute(calls2); 

token.mint(100, bob); // does not use the delegation
```

Is that how you envisage the logic to work, or would you prefer to explicitly revoke the delegation after the first transaction? Let me know if I'm missing something as well.

klkvr replied:
if an authorization is applied to transaction, then it is persisted in database and there's no need to keep this delegation in `TxEnv` or to attach to any other transactions

storing it in TxEnv for all other transactions would result in gas cost for all subsequent calls including the cost of authorization which is not something we want

evchip replied:
Got it, thanks. I've added logic to set `authorization_list` to `None` if `active_delegation` is `None`, and added a test to assert that attaching another delegation to a tx in the same bundle still switches out the bytecode on the EOA.

evchip replied:
I misunderstood - we need to clear `TxEnv.authorization_list` between transactions. would adding `unset_delegation()` to clear it when `tx.authorization_list()` is `None` work? Or is there a better approach?

klkvr replied:
imo we could just handle this differently, similar to how to/data/value are set, without touching the global executor env.

`ScriptRunner::simulate` could take authorization list as an additional argument, and then we can either
1. Manually call `Executor::build_test_env`, set authorization list to the created temporary TxEnv, and pass it to `call_with_env` just as `Executor::call_raw` does
2. Or add a helper method `Executor::call_raw_with_authorization` which would be same as `call_raw` but also accept authorization list

evchip replied:
Got it, thanks for the clarification. I've pushed [a commit](https://github.com/foundry-rs/foundry/pull/9236/commits/2a30b429a2fa8f2a9e32277374f2bcb4502a2c09) which uses your option 2.

evchip commented on crates/cheatcodes/src/script.rs:
```diff
@@ -29,6 +32,65 @@ impl Cheatcode for broadcast_2Call {
     }
 }
 
+impl Cheatcode for attachDelegationCall {
+    fn apply_stateful(&self, ccx: &mut CheatsCtxt) -> Result {
+        let Self { implementation, authority, v, r, s } = self;
```
I'm not seeing how to construct the authorization without the authority's nonce, which we'd need to recover the authority from the signature. Can you clarify the flow you're thinking of?

klkvr replied:
we can get the nonce from state

evchip replied:
Right, but to get the nonce from state we need the authority's address to load their account - am I missing some other way to get their nonce?

klkvr replied:
in `signDelegation` we can get authority address from private key, then get nonce, construct and sign delegation

and in `attachDelegation` we can just recover authority from the signed delegation. right now the passed authority address is anyway expected to be the same as recovered one making it redundant

klkvr replied:
ah I see what you mean. didn't notice that `attachDelegation` doesn't accept `nonce`

let's make `signDelegation` return a new `struct SignedDelegation` which will contain `v,r,s,nonce,implementation`. And then `attachDelegation` would just accept this object

evchip replied:
I've implemented a new cheatcode `signAndAttachDelegation` and refactored the existing cheatcodes to use `struct SignedDelegation`. Let me know what you think. Thanks!

klkvr reviewed (approved):
nice, lgtm!

yash-atreya reviewed (approved):
Can be merged @grandizzy
